resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala

```diff
@@ -16,256 +16,130 @@
-  private val driverLimitCores = sparkConf.getOption(KUBERNETES_DRIVER_LIMIT_CORES.key)
 
-  // Memory settings
-  private val driverMemoryMb = sparkConf.get(org.apache.spark.internal.config.DRIVER_MEMORY)
-  private val memoryOverheadMb = sparkConf
-    .get(KUBERNETES_DRIVER_MEMORY_OVERHEAD)
-    .getOrElse(math.max((MEMORY_OVERHEAD_FACTOR * driverMemoryMb).toInt,
-      MEMORY_OVERHEAD_MIN))
-  private val driverContainerMemoryWithOverhead = driverMemoryMb + memoryOverheadMb
-  private val customLabels = sparkConf.get(KUBERNETES_DRIVER_LABELS)
-  private val customAnnotations = sparkConf.get(KUBERNETES_DRIVER_ANNOTATIONS)
-
-  private val driverExtraClasspath = sparkConf.get(
-    org.apache.spark.internal.config.DRIVER_CLASS_PATH)
-  private val driverJavaOptions = sparkConf.get(
+  private val driverJavaOptions = submissionSparkConf.get(
     org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
 
+   /**
+    * Run command that initalizes a DriverSpec that will be updated
+    * after each KubernetesSubmissionStep in the sequence that is passed in.
+    * The final driver-spec will be used to build the Driver Container,
+    * Driver Pod, and Kubernetes Resources
+    *
+    */
   def run(): Unit = {
-    val arguments = (pythonResource map {p => p.arguments}).getOrElse(appArgs)
-    val driverCustomLabels = ConfigurationUtils.combinePrefixedKeyValuePairsWithDeprecatedConf(
-      sparkConf,
-      KUBERNETES_DRIVER_LABEL_PREFIX,
-      KUBERNETES_DRIVER_LABELS,
-      "label")
-    require(!driverCustomLabels.contains(SPARK_APP_ID_LABEL), s"Label with key " +
-        s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
-        s" operations.")
-
-    val driverCustomAnnotations = ConfigurationUtils.combinePrefixedKeyValuePairsWithDeprecatedConf(
-      sparkConf,
-      KUBERNETES_DRIVER_ANNOTATION_PREFIX,
-      KUBERNETES_DRIVER_ANNOTATIONS,
-      "annotation")
-    require(!driverCustomAnnotations.contains(SPARK_APP_NAME_ANNOTATION),
-        s"Annotation with key $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for" +
-        s" Spark bookkeeping operations.")
-    val allDriverLabels = driverCustomLabels ++ Map(
-        SPARK_APP_ID_LABEL -> kubernetesAppId,
-        SPARK_ROLE_LABEL -> SPARK_POD_DRIVER_ROLE)
-
-    val driverExtraClasspathEnv = driverExtraClasspath.map { classPath =>
-      new EnvVarBuilder()
-        .withName(ENV_SUBMIT_EXTRA_CLASSPATH)
-        .withValue(classPath)
-        .build()
+    // Set new metadata and a new spec so that submission steps can use PodBuilder#editMetadata()
+    // and/or PodBuilder#editSpec() safely.
+    val basePod = new PodBuilder().withNewMetadata().endMetadata().withNewSpec().endSpec().build()
+    var currentDriverSpec = KubernetesDriverSpec(
+      driverPod = basePod,
+      driverContainer = new ContainerBuilder().build(),
+      driverSparkConf = submissionSparkConf.clone(),
```

Comment:
I think we want to enforce that all submission steps would specify all of the fields, so I don't think we want to create the shortcut.